vpr/src/base/read_place.cpp

```diff
@@ -97,194 +97,129 @@ void read_place(const char* net_file,
+            if (!is_place_file) {
+                if (place_ctx.block_locs[blk_id].loc.x != OPEN) {
+                    VPR_THROW(VPR_ERROR_PLACE, "The block with ID %d is listed twice in the constraints file.\n", blk_id);
+                }
+            }
 
-        if (k >= physical_tile->capacity || k < 0) {
-            VPR_THROW(VPR_ERROR_PLACE_F, constraints_file, vtr::get_file_line_number_of_last_opened_file(),
-                      "Block %s subtile number (%d) is out of range.\n", bname, k);
-        }
-        place_ctx.grid_blocks[i][j].blocks[k] = bnum;
-        place_ctx.grid_blocks[i][j].usage++;
+            //Check if block location is out of range of grid dimensions
+            if (!is_place_file) {
+                if (block_x < 0 || block_x > int(device_ctx.grid.width() - 1)
+                    || block_y < 0 || block_y > int(device_ctx.grid.height() - 1)) {
+                    VPR_THROW(VPR_ERROR_PLACE, "The block with ID %d is out of range at location (%d, %d). \n", blk_id, block_x, block_y);
+                }
+            }
 
-        constrained_blocks.insert(bnum);
+            if (place_ctx.block_locs.size() != cluster_ctx.clb_nlist.blocks().size()) {
+                //Resize if needed
+                place_ctx.block_locs.resize(cluster_ctx.clb_nlist.blocks().size());
+            }
 
-        ptr = vtr::fgets(buf, vtr::bufsize, fp);
-    }
+            //Set the location
+            place_ctx.block_locs[blk_id].loc.x = block_x;
+            place_ctx.block_locs[blk_id].loc.y = block_y;
+            place_ctx.block_locs[blk_id].loc.sub_tile = sub_tile_index;
 
-    for (auto blk_id : cluster_ctx.clb_nlist.blocks()) {
-        auto result = constrained_blocks.find(blk_id);
-        if (result == constrained_blocks.end()) {
-            continue;
-        }
+            //need to lock down blocks  and mark grid block usage if it is a constraints file
+            if (!is_place_file) {
+                place_ctx.block_locs[blk_id].is_fixed = true;
+                place_ctx.grid_blocks[block_x][block_y].blocks[sub_tile_index] = blk_id;
+                place_ctx.grid_blocks[block_x][block_y].usage++;
+            }
 
-        if (place_ctx.block_locs[blk_id].loc.x == OPEN) {
-            VPR_THROW(VPR_ERROR_PLACE_F, constraints_file, 0,
-                      "Block %s location was not specified in the constraints file.\n", cluster_ctx.clb_nlist.block_name(blk_id).c_str());
+        } else {
+            //Unrecognized
+            vpr_throw(VPR_ERROR_PLACE_F, place_file, lineno,
+                      "Invalid line '%s' in placement file",
+                      line.c_str());
         }
     }
 
-    fclose(fp);
-    free_hash_table(hash_table);
-    VTR_LOG("Successfully read %s.\n", constraints_file);
-    VTR_LOG("\n");
+    if (!is_place_file) {
+        VTR_LOG("Successfully read %s.\n", place_file);
+        VTR_LOG("\n");
+    }
+    place_ctx.placement_id = vtr::secure_digest_file(place_file);
```

Comment:
This is a bug; shouldn't generate the signature (used to verify that a routing file uses a compatible placement if we later read in that routing file) for the placement if we're not reading a placement file.